core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLRecordReader.java

```diff
@@ -0,0 +1,332 @@
+            }
+            break;
+          case XMLStreamConstants.END_ELEMENT:
+            String nodeNameEnd = reader.getLocalName();
+            if (nodeFound) {
+              //add closing tag
+              xmlRecord.append(OPENING_END_TAG_DELIMITER).append(nodeNameEnd).append(CLOSING_END_TAG_DELIMITER);
+              if (nodeNameEnd.equals(lastNode)) {
+                nodeFound = false;
+                //set flag for XML record is ready to emit.
+                xmlRecordReady = true;
+              }
+            }
+            nodeLevel--;
+            break;
+          case XMLStreamConstants.START_DOCUMENT:
+            break;
+        }
+        if (xmlRecordReady) {
+          currentValue.put(fileName, xmlRecord.toString());
+          return true;
+        }
+      }
+    } catch (XMLStreamException exception) {
+      throw new IllegalArgumentException(exception.getMessage());
+    }
+    actionsAfterEOF();
+    return false;
+  }
+
+  /**
+   * Method to append start tag information along with attributes if any
+   * @param nodeNameStart
+   * @param xmlRecord
+   */
+  private void appendStartTagInformation(String nodeNameStart, StringBuilder xmlRecord) {
+    xmlRecord.append(OPENING_START_TAG_DELIMITER).append(nodeNameStart);
+    int count = reader.getAttributeCount();
+    for (int i = 0; i < count; i++) {
+      xmlRecord.append(" ").append(reader.getAttributeLocalName(i)).append("=\"").append(reader.getAttributeValue(i))
+        .append("\"");
+    }
+    xmlRecord.append(CLOSING_START_TAG_DELIMITER);
+  }
+
+  /**
+   * Method to take actions after EOF reached.
+   * @throws IOException
+   */
+  private void actionsAfterEOF() throws IOException {
+    if (StringUtils.isNotEmpty(fileAction)) {
+      processFileAction();
+    }
+    updateFileTrackingInfo();
+  }
+
+  /**
+   * Method to process file with actions (Delete, Move, Archive ) specified.
+   * @throws IOException
+   */
+  private void processFileAction() throws IOException {
+    fileAction = fileAction.toLowerCase();
+    try {
+      switch (fileAction) {
+        case "delete":
+          fs.delete(file, true);
+          break;
+        case "move":
+          Path tagetFileMovePath = new Path(targetFolder + file.getName());
+          fs.rename(file, tagetFileMovePath);
+          break;
+        case "archive":
+          FileOutputStream archivedStream = new FileOutputStream(targetFolder + file.getName() + ".zip");
+          ZipOutputStream zipArchivedStream = new ZipOutputStream(archivedStream);
+          FSDataInputStream fdDataInputStream = fs.open(file);
+          zipArchivedStream.putNextEntry(new ZipEntry(file.getName()));
+          int length;
+          byte[] buffer = new byte[1024];
+          while ((length = fdDataInputStream.read(buffer)) > 0) {
+            zipArchivedStream.write(buffer, 0, length);
+          }
+          zipArchivedStream.closeEntry();
+          fdDataInputStream.close();
```

Comment:
please use try with resources. If there is an exception between line 261 and 267, ``fdDataInputStream``, ``zipArchivedStream`` won't be closed